syntax = "proto3";

message Message {
  string sender = 1;
  string receiver = 2;
  string content = 3;
  int64 timestamp = 4;
}

message Chat {
  string name = 1;
  repeated Message messages = 2;
}

message CreateChatRequest {
  string username = 1;
}

message SendMessageRequest {
  string sender = 1;
  string receiver = 2;
  string content = 3;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
}

message RegisterRequest {
  string username = 1;
  string password = 2;
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
}

service Messenger {
  rpc SendMessage(SendMessageRequest) returns (Message);
  rpc GetChatHistory(Chat) returns (stream Message);
  rpc CreateChat(CreateChatRequest) returns (Chat);
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Register(RegisterRequest) returns (RegisterResponse);
}
